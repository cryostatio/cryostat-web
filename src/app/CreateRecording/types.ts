/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { EventTemplate, KeyValue } from '@app/Shared/Services/api.types';
import { ValidatedOptions } from '@patternfly/react-core';

export type EventTemplateIdentifier = Pick<EventTemplate, 'name' | 'type'>;

interface _FormBaseData {
  name: string;
  template?: EventTemplateIdentifier;
  labels: KeyValue[];
  continuous: boolean;
  archiveOnStop: boolean;
  restart: boolean;
  duration: number;
  durationUnit: number;
  skipDurationCheck?: boolean;
  maxAge: number;
  maxAgeUnit: number;
  maxSize: number;
  maxSizeUnit: number;
  toDisk: boolean;
}

interface _FormValidationData {
  nameValid: ValidatedOptions;
  labelsValid: ValidatedOptions;
  durationValid: ValidatedOptions;
}

export type CustomRecordingFormData = _FormBaseData & _FormValidationData;
