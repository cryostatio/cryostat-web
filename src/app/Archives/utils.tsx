/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { RecordingDirectory, Target } from '@app/Shared/Services/api.types';

export const includesDirectory = (arr: RecordingDirectory[], dir: RecordingDirectory): boolean => {
  return arr.some((t) => t.connectUrl === dir.connectUrl);
};

export const indexOfDirectory = (arr: RecordingDirectory[], dir: RecordingDirectory): number => {
  let index = -1;
  arr.forEach((d, idx) => {
    if (d.connectUrl === dir.connectUrl) {
      index = idx;
    }
  });
  return index;
};

export const getTargetFromDirectory = (dir: RecordingDirectory): Target => {
  return {
    connectUrl: dir.connectUrl,
    alias: dir.jvmId,
    labels: [],
    annotations: {
      cryostat: [],
      platform: [],
    },
  };
};
