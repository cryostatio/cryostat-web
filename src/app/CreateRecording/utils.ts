/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { TemplateType } from '@app/Shared/Services/api.types';
import { EventTemplateIdentifier } from './types';

export const RecordingNamePattern = /^[\w_]+$/;

export const DurationPattern = /^[1-9][0-9]*$/;

export const isRecordingNameValid = (name: string) => RecordingNamePattern.test(name);

export const isDurationValid = (duration: number) => DurationPattern.test(`${duration}`);

export const templateFromEventSpecifier = (eventSpecifier: string): EventTemplateIdentifier | undefined => {
  const regex = /^template=([a-zA-Z0-9]+)(?:,type=([a-zA-Z0-9]+))?$/im;
  if (eventSpecifier && regex.test(eventSpecifier)) {
    const parts = regex.exec(eventSpecifier);
    if (parts) {
      return {
        name: parts[1],
        type: parts[2] as TemplateType,
      };
    }
  }
  return undefined;
};
