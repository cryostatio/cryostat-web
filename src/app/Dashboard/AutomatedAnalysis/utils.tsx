/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { AnalysisResult, Evaluation } from '@app/Shared/Services/api.types';
import { Stack, StackItem, Label, Title, Text } from '@patternfly/react-core';
import _ from 'lodash';

export const transformAADescription = (result: AnalysisResult): JSX.Element => {
  const format = (s: Evaluation): JSX.Element => {
    if (typeof s === 'string') {
      return <Text>{s}</Text>;
    }
    if (Array.isArray(s)) {
      return (
        <Stack>
          {s.map((e) => (
            <StackItem key={e.setting}>
              <Title headingLevel={'h6'}>{e.setting}</Title>
              <Label>
                {e.name}={e.value}
              </Label>
            </StackItem>
          ))}
        </Stack>
      );
    }
    throw `Unrecognized item: ${s}`;
  };
  return (
    <div>
      {Object.entries(result.evaluation || {}).map(([k, v]) =>
        v && v.length ? (
          <div key={k}>
            <span>
              <Title headingLevel={'h5'}>{_.capitalize(k)}</Title>
              {format(result.evaluation[k])}
            </span>
            <br />
          </div>
        ) : (
          <div key={k}></div>
        ),
      )}
    </div>
  );
};
