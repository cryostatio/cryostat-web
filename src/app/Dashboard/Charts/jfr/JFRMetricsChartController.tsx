/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ApiService, RecordingState } from '@app/Shared/Services/Api.service';
import { NotificationCategory, NotificationChannel } from '@app/Shared/Services/NotificationChannel.service';
import { SettingsService } from '@app/Shared/Services/Settings.service';
import { NO_TARGET, Target, TargetService } from '@app/Shared/Services/Target.service';
import {
  BehaviorSubject,
  concatMap,
  distinctUntilChanged,
  finalize,
  first,
  map,
  merge,
  Observable,
  of,
  pairwise,
  ReplaySubject,
  Subscription,
  switchMap,
  tap,
  throttleTime,
} from 'rxjs';

export const RECORDING_NAME = 'dashboard_metrics';

export enum ControllerState {
  UNKNOWN = 0,
  NO_DATA = 1,
  READY = 2,
}

export class JFRMetricsChartController {
  private readonly _state$ = new BehaviorSubject<ControllerState>(ControllerState.UNKNOWN);
  private readonly _refCount$ = new BehaviorSubject<number>(0);
  private readonly _updates$ = new ReplaySubject<void>(1);
  private readonly _lazy: Subscription;
  private _attach: Subscription | undefined;

  constructor(
    private readonly _api: ApiService,
    private readonly _target: TargetService,
    private readonly _notifications: NotificationChannel,
    private readonly _settings: SettingsService,
  ) {
    this._lazy = this._refCount$
      .pipe(
        map((v) => v > 0),
        pairwise(),
      )
      .subscribe(([prev, curr]) => {
        if (!prev && curr) {
          // first subscriber joined
          this._start();
        }
        if (prev && !curr) {
          // last subscriber left
          this._stop();
        }
      });
  }

  attach(): Observable<ControllerState> {
    this._refCount$.next(this._refCount$.value + 1);
    return this._state$.asObservable().pipe(
      distinctUntilChanged(),
      finalize(() => this._refCount$.next(this._refCount$.value - 1)),
    );
  }

  requestRefresh(): void {
    this._updates$.next();
  }

  _tearDown() {
    this._state$.next(ControllerState.UNKNOWN);
    this._lazy.unsubscribe();
    this._stop();
  }

  private _stop(): void {
    if (this._attach) {
      this._attach.unsubscribe();
      this._attach = undefined;
    }
  }

  private _start(): void {
    this._stop();
    this._attach = merge(
      merge(
        this._updates$.pipe(throttleTime(this._settings.chartControllerConfig().minRefresh * 1000)),
        this._notifications.messages(NotificationCategory.ActiveRecordingCreated),
        this._notifications.messages(NotificationCategory.ActiveRecordingDeleted),
        this._notifications.messages(NotificationCategory.ActiveRecordingStopped),
      ).pipe(switchMap((_) => this._target.target().pipe(first()))),
      this._target.target().pipe(tap((_) => this._state$.next(ControllerState.UNKNOWN))),
    )
      .pipe(concatMap((t) => this._hasRecording(t)))
      .subscribe((v) => {
        this._state$.next(v ? ControllerState.READY : ControllerState.NO_DATA);
        if (v) {
          this._api
            .uploadActiveRecordingToGrafana(RECORDING_NAME)
            .pipe(first())
            .subscribe((_) => {
              this._state$.next(ControllerState.READY);
            });
        }
      });
  }

  private _hasRecording(target: Target): Observable<boolean> {
    if (target === NO_TARGET) {
      return of(false);
    }
    return this._api.targetHasRecording(target, {
      state: RecordingState.RUNNING,
      labels: [`origin=${RECORDING_NAME}`],
    });
  }
}
