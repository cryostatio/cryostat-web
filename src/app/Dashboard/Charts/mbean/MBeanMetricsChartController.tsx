/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ApiService } from '@app/Shared/Services/Api.service';
import { MBeanMetrics, Target } from '@app/Shared/Services/api.types';
import { SettingsService } from '@app/Shared/Services/Settings.service';
import { TargetService } from '@app/Shared/Services/Target.service';
import {
  BehaviorSubject,
  catchError,
  concatMap,
  distinctUntilChanged,
  finalize,
  first,
  map,
  merge,
  Observable,
  of,
  pairwise,
  ReplaySubject,
  Subject,
  Subscription,
  tap,
  throttleTime,
} from 'rxjs';

export class MBeanMetricsChartController {
  private readonly _metrics = new Map<string, string[]>();
  private readonly _state$ = new Subject<MBeanMetrics>();
  private readonly _loading$ = new BehaviorSubject<boolean>(true);
  private readonly _refCount$ = new BehaviorSubject<number>(0);
  private readonly _updates$ = new ReplaySubject<void>(1);
  private readonly _lazy: Subscription;
  private _attach: Subscription | undefined;

  constructor(
    private readonly _api: ApiService,
    private readonly _target: TargetService,
    private readonly _settings: SettingsService,
  ) {
    this._lazy = this._refCount$
      .pipe(
        map((v) => v > 0),
        pairwise(),
      )
      .subscribe(([prev, curr]) => {
        if (!prev && curr) {
          // first subscriber joined
          this._start();
        }
        if (prev && !curr) {
          // last subscriber left
          this._stop();
        }
      });
  }

  attach(category: string, fields: string[]): Observable<MBeanMetrics> {
    this._refCount$.next(this._refCount$.value + 1);
    if (!this._metrics.has(category)) {
      this._metrics.set(category, []);
    }
    this._metrics.get(category)?.push(...fields);
    return this._state$.asObservable().pipe(
      distinctUntilChanged(),
      finalize(() => {
        this._refCount$.next(this._refCount$.value - 1);
        const original = this._metrics.get(category) || [];
        const updated = [...original];
        fields.forEach((field) => {
          const idx = updated.findIndex((e) => e === field);
          if (idx >= 0) {
            updated.splice(idx, 1);
          }
        });
        this._metrics.get(category)?.push(...updated);
      }),
    );
  }

  requestRefresh(): void {
    this._updates$.next();
  }

  loading(): Observable<boolean> {
    return this._loading$.asObservable().pipe(distinctUntilChanged());
  }

  _tearDown() {
    this._state$.next({});
    this._lazy.unsubscribe();
    this._stop();
  }

  private _stop(): void {
    if (this._attach) {
      this._attach.unsubscribe();
      this._attach = undefined;
    }
  }

  private _start(): void {
    this._stop();
    this._attach = merge(
      this._updates$.pipe(
        throttleTime(this._settings.chartControllerConfig().minRefresh * 1000),
        concatMap((_) => this._target.target().pipe(first())),
      ),
      this._target.target(),
    )
      .pipe(
        tap((_) => this._loading$.next(true)),
        concatMap((t) => this._queryMetrics(t)),
        catchError((_) => of({})),
        tap((_) => this._loading$.next(false)),
      )
      .subscribe((v) => this._state$.next(v));
  }

  private _queryMetrics(target?: Target): Observable<MBeanMetrics> {
    if (!target) {
      return of({});
    }
    const q: string[] = [];
    const m = new Map<string, Set<string>>();
    this._metrics.forEach((fields, category) => {
      const s = m.get(category) || new Set<string>();
      fields.forEach((f) => s.add(f));
      m.set(category, s);
    });
    m.forEach((s, k) => {
      let l = `${k} {`;
      s.forEach((f) => (l += `\t${f}\n`));
      l += '}';
      q.push(l);
    });
    if (q.length === 0) {
      return of({});
    }
    return this._api.getTargetMBeanMetrics(target, q);
  }
}
