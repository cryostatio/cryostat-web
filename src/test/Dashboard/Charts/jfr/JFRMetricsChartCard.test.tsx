/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

jest.mock('@app/Dashboard/Charts/jfr/JFRMetricsChartController');

import { ChartContext } from '@app/Dashboard/Charts/context';
import { JFRMetricsChartCard, kindToId } from '@app/Dashboard/Charts/jfr/JFRMetricsChartCard';
import { JFRMetricsChartController, ControllerState } from '@app/Dashboard/Charts/jfr/JFRMetricsChartController';
import { MBeanMetricsChartController } from '@app/Dashboard/Charts/mbean/MBeanMetricsChartController';
import { ThemeSetting } from '@app/Settings/types';
import { defaultServices } from '@app/Shared/Services/Services';
import { cleanup, screen } from '@testing-library/react';
import { of } from 'rxjs';
import { mockMediaQueryList, render, renderSnapshot } from '../../../utils';

const mockDashboardUrl = 'http://localhost:3000';
jest.spyOn(defaultServices.api, 'grafanaDashboardUrl').mockReturnValue(of(mockDashboardUrl));

const mockTarget = {
  connectUrl: 'service:jmx:rmi://someUrl',
  alias: 'fooTarget',
  jvmId: 'foo',
  labels: [],
  annotations: { cryostat: [], platform: [] },
};
jest.spyOn(defaultServices.target, 'target').mockReturnValue(of(mockTarget));

jest.spyOn(defaultServices.settings, 'themeSetting').mockReturnValue(of(ThemeSetting.LIGHT));
jest.spyOn(defaultServices.settings, 'media').mockReturnValue(of(mockMediaQueryList));

const mockJfrController = new JFRMetricsChartController(
  defaultServices.api,
  defaultServices.target,
  defaultServices.notificationChannel,
  defaultServices.settings,
);
const mockMbeanController = new MBeanMetricsChartController(
  defaultServices.api,
  defaultServices.target,
  defaultServices.settings,
);
const mockChartContext = {
  jfrController: mockJfrController,
  mbeanController: mockMbeanController,
};

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}));

describe('<JFRMetricsChartCard />', () => {
  beforeEach(() => {
    jest.mocked(mockNavigate).mockClear();
  });
  afterEach(cleanup);

  it('renders correctly', async () => {
    jest.spyOn(mockJfrController, 'attach').mockReturnValue(of(ControllerState.READY));

    const tree = await renderSnapshot({
      routerConfigs: {
        routes: [
          {
            path: '/',
            element: <JFRMetricsChartCard chartKind={'CPU Load'} duration={120} period={10} span={6} dashboardId={0} />,
          },
        ],
      },
      providers: [{ kind: ChartContext.Provider, instance: mockChartContext }],
    });
    expect(tree?.toJSON()).toMatchSnapshot();
  });

  it('renders loading state correctly', async () => {
    jest.spyOn(mockJfrController, 'attach').mockReturnValue(of(ControllerState.UNKNOWN));

    const tree = await renderSnapshot({
      routerConfigs: {
        routes: [
          {
            path: '/',
            element: <JFRMetricsChartCard chartKind={'CPU Load'} duration={120} period={10} span={6} dashboardId={0} />,
          },
        ],
      },
      providers: [{ kind: ChartContext.Provider, instance: mockChartContext }],
    });
    expect(tree?.toJSON()).toMatchSnapshot();
  });

  it('renders empty state correctly', async () => {
    jest.spyOn(mockJfrController, 'attach').mockReturnValue(of(ControllerState.NO_DATA));

    const tree = await renderSnapshot({
      routerConfigs: {
        routes: [
          {
            path: '/',
            element: <JFRMetricsChartCard chartKind={'CPU Load'} duration={120} period={10} span={6} dashboardId={0} />,
          },
        ],
      },
      providers: [{ kind: ChartContext.Provider, instance: mockChartContext }],
    });
    expect(tree?.toJSON()).toMatchSnapshot();
  });

  it('renders empty state with information and action button', async () => {
    jest.spyOn(mockJfrController, 'attach').mockReturnValue(of(ControllerState.NO_DATA));

    render({
      routerConfigs: {
        routes: [
          {
            path: '/',
            element: <JFRMetricsChartCard chartKind={'CPU Load'} duration={120} period={10} span={6} dashboardId={0} />,
          },
        ],
      },
      providers: [{ kind: ChartContext.Provider, instance: mockChartContext }],
    });

    expect(screen.getByText('CPU Load (last 120s, every 10s)')).toBeInTheDocument();
    expect(screen.getByText('No source recording')).toBeInTheDocument();
    expect(
      screen.getByText((s) => s.includes('Metrics cards display data taken from running flight recordings')),
    ).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /create/i })).toBeInTheDocument();
  });

  it('navigates to recording creation with prefilled state when empty state button clicked', async () => {
    jest.spyOn(mockJfrController, 'attach').mockReturnValue(of(ControllerState.NO_DATA));

    const { user } = render({
      routerConfigs: {
        routes: [
          {
            path: '/',
            element: <JFRMetricsChartCard chartKind={'CPU Load'} duration={120} period={10} span={6} dashboardId={0} />,
          },
        ],
      },
      providers: [{ kind: ChartContext.Provider, instance: mockChartContext }],
    });

    await user.click(screen.getByRole('button', { name: /create/i }));
    expect(mockNavigate).toHaveBeenCalledWith('/recordings/create', {
      state: {
        name: 'dashboard_metrics',
        template: {
          name: 'Continuous',
          type: 'TARGET',
        },
        restart: true,
        labels: [{ key: 'origin', value: 'dashboard_metrics' }],
        duration: -1,
        skipDurationCheck: true,
        maxAge: 120,
        maxAgeUnit: 1,
        maxSize: 100 * 1024 * 1024,
        maxSizeUnit: 1,
      },
    });
  });

  it.each([
    ['CPU Load', 120, 5],
    ['Heap Usage', 90, 10],
    ['Network Utilization', 60, 15],
    ['File I/O', 30, 20],
  ])('renders iframe', async (chartKind: string, duration: number, period: number) => {
    jest.spyOn(mockJfrController, 'attach').mockReturnValue(of(ControllerState.READY));

    const { container } = render({
      routerConfigs: {
        routes: [
          {
            path: '/',
            element: (
              <JFRMetricsChartCard chartKind={chartKind} duration={duration} period={period} span={6} dashboardId={0} />
            ),
          },
        ],
      },
      providers: [{ kind: ChartContext.Provider, instance: mockChartContext }],
    });

    const iframe = container.querySelector('iframe');
    expect(iframe).toBeTruthy();
    const u = new URL(iframe?.src || '');
    expect(u.host).toBe('localhost:3000');
    expect(u.protocol).toBe('http:');
    const params = new URLSearchParams();
    params.set('theme', 'light');
    params.set('panelId', String(kindToId(chartKind)));
    params.set('to', 'now');
    params.set('from', `now-${duration}s`);
    params.set('refresh', `${period}s`);
    expect(u.searchParams.toString()).toEqual(params.toString());
  });
});
