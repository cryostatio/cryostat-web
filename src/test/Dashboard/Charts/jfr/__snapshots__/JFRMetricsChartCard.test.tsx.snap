// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<JFRMetricsChartCard /> renders correctly 1`] = `
<div
  className="draggable-ref-wrapper"
  onDragStart={[Function]}
  onTransitionEnd={[Function]}
  style={Object {}}
>
  <div
    className="dashboard-card-resizable-wrapper"
  >
    <div
      className="pf-v5-c-card pf-m-compact pf-m-rounded dashboard-card"
      data-ouia-component-id="OUIA-Generated-Card-1"
      data-ouia-component-type="PF5/Card"
      data-ouia-safe={true}
      id="CPU Load-chart-card"
      style={
        Object {
          "height": "380px",
        }
      }
      title="CPU Load"
    >
      <div
        className="draggable-ref__grip"
        data-quickstart-id="card-draggable-grip"
        draggable={true}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="pf-v5-c-card__header"
        >
          <div
            className="pf-v5-c-card__actions"
          >
            <button
              aria-disabled={false}
              aria-label="Synchronize CPU Load chart"
              className="pf-v5-c-button pf-m-plain"
              data-ouia-component-id="OUIA-Generated-Button-plain-1"
              data-ouia-component-type="PF5/Button"
              data-ouia-safe={true}
              disabled={false}
              onClick={[Function]}
              role={null}
              type="button"
            >
              <svg
                aria-hidden={true}
                aria-labelledby={null}
                className="pf-v5-svg"
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 512 512"
                width="1em"
              >
                <path
                  d="M370.72 133.28C339.458 104.008 298.888 87.962 255.848 88c-77.458.068-144.328 53.178-162.791 126.85-1.344 5.363-6.122 9.15-11.651 9.15H24.103c-7.498 0-13.194-6.807-11.807-14.176C33.933 94.924 134.813 8 256 8c66.448 0 126.791 26.136 171.315 68.685L463.03 40.97C478.149 25.851 504 36.559 504 57.941V192c0 13.255-10.745 24-24 24H345.941c-21.382 0-32.09-25.851-16.971-40.971l41.75-41.749zM32 296h134.059c21.382 0 32.09 25.851 16.971 40.971l-41.75 41.75c31.262 29.273 71.835 45.319 114.876 45.28 77.418-.07 144.315-53.144 162.787-126.849 1.344-5.363 6.122-9.15 11.651-9.15h57.304c7.498 0 13.194 6.807 11.807 14.176C478.067 417.076 377.187 504 256 504c-66.448 0-126.791-26.136-171.315-68.685L48.97 471.03C33.851 486.149 8 475.441 8 454.059V320c0-13.255 10.745-24 24-24z"
                />
              </svg>
            </button>
            <button
              aria-disabled={false}
              aria-label="Pop out CPU Load chart"
              className="pf-v5-c-button pf-m-plain"
              data-ouia-component-id="OUIA-Generated-Button-plain-2"
              data-ouia-component-type="PF5/Button"
              data-ouia-safe={true}
              disabled={false}
              onClick={[Function]}
              role={null}
              type="button"
            >
              <svg
                aria-hidden={true}
                aria-labelledby={null}
                className="pf-v5-svg"
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 512 512"
                width="1em"
              >
                <path
                  d="M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z"
                />
              </svg>
            </button>
          </div>
          <div
            className="pf-v5-c-card__header-main"
          >
            <div
              className="pf-v5-c-card__title"
            >
              <div
                className="pf-v5-c-card__title-text"
                id="CPU Load-chart-card-title"
              >
                CPU Load (last 120s, every 10s)
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="pf-v5-c-card__body"
      >
        <div
          className="grafana-iframe-wrapper"
        >
          <iframe
            className="disabled-pointer"
            src="http://localhost:3000/d-solo/main?theme=light&panelId=3&to=now&from=now-120s&refresh=10s"
          />
        </div>
      </div>
    </div>
    <div
      className="resizable-ref"
      onMouseDown={[Function]}
    />
  </div>
</div>
`;

exports[`<JFRMetricsChartCard /> renders empty state correctly 1`] = `
<div
  className="draggable-ref-wrapper"
  onDragStart={[Function]}
  onTransitionEnd={[Function]}
  style={Object {}}
>
  <div
    className="dashboard-card-resizable-wrapper"
  >
    <div
      className="pf-v5-c-card pf-m-compact pf-m-rounded dashboard-card"
      data-ouia-component-id="OUIA-Generated-Card-3"
      data-ouia-component-type="PF5/Card"
      data-ouia-safe={true}
      id="CPU Load-chart-card"
      style={Object {}}
      title="CPU Load"
    >
      <div
        className="draggable-ref__grip"
        data-quickstart-id="card-draggable-grip"
        draggable={true}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="pf-v5-c-card__header"
        >
          <div
            className="pf-v5-c-card__actions"
          />
          <div
            className="pf-v5-c-card__header-main"
          >
            <div
              className="pf-v5-c-card__title"
            >
              <div
                className="pf-v5-c-card__title-text"
                id="CPU Load-chart-card-title"
              >
                CPU Load (last 120s, every 10s)
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="pf-v5-c-card__body"
      >
        <div
          className="pf-v5-l-bullseye"
        >
          <div
            className="pf-v5-c-empty-state pf-m-lg"
          >
            <div
              className="pf-v5-c-empty-state__content"
            >
              <div
                className="pf-v5-c-empty-state__header"
              >
                <div
                  className="pf-v5-c-empty-state__icon"
                >
                  <svg
                    aria-hidden={true}
                    aria-labelledby={null}
                    className="pf-v5-svg"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    viewBox="0 0 896 1024"
                    width="1em"
                  >
                    <path
                      d="M750.6252,63 L896.0002,223 L750.6252,383 L704.0002,339.594 L780.807,255.06 L179.0074,254.909 L178.9774,254.909 C147.0174,254.909 117.0004,268.239 94.4554,292.446 C71.9454,316.615 59.6154,348.704 59.7374,382.802 C59.9914,453.334 113.4824,511 178.9784,511 L717.0204,511 C764.8314,511 809.7404,530.399 843.4754,566.14 C877.3464,602.026 896.0004,649.93 896.0004,701.029 L896.0004,701.689 C896.0004,752.688 877.3844,800.818 843.5834,837.211 C809.9954,873.375 765.2134,893.895 717.4864,894.992 L717.1364,895 L276.310151,895.001486 C261.798404,956.828303 206.189494,1003 140,1003 C62.804,1003 0,940.196 0,863 C0,785.804 62.804,723 140,723 C206.189852,723 261.799004,769.172196 276.310386,830.999515 L716.5394,831 C748.5984,830.173 778.6564,816.448 801.2074,792.166 C823.8134,767.826 836.2624,735.694 836.2624,701.689 L836.2624,701.029 C836.2624,667.119 823.8964,635.341 801.4404,611.55 C778.9654,587.739 748.9844,575 717.0204,575 L178.9784,575 C80.6694,575 0.0014,489 0.0014,383.047 C0.0014,332 18.3334,283.634 52.1394,247.337 C85.9744,211.008 131.0224,191 178.9774,191 L179.0224,191 L781.009,191.162 L704.0002,106.406 L750.6252,63 Z M140,787 C98.093,787 64,821.094 64,863 C64,904.906 98.093,939 140,939 C181.907,939 216,904.906 216,863 C216,821.094 181.907,787 140,787 Z"
                    />
                  </svg>
                </div>
                <div
                  className="pf-v5-c-empty-state__title"
                >
                  <h2
                    className="pf-v5-c-empty-state__title-text"
                  >
                    No source Recording
                  </h2>
                </div>
              </div>
              <div
                className="pf-v5-c-empty-state__body"
              >
                Metrics cards display data taken from running Flight Recordings with the label 
                <span
                  className="pf-v5-c-label pf-m-blue pf-m-compact"
                >
                  <span
                    className="pf-v5-c-label__content"
                  >
                    <span
                      className="pf-v5-c-label__text"
                    >
                      origin=dashboard_metrics
                    </span>
                  </span>
                </span>
                . No such Recordings are currently available.
              </div>
              <div
                className="pf-v5-c-empty-state__footer"
              >
                <button
                  aria-disabled={false}
                  aria-label={null}
                  className="pf-v5-c-button pf-m-primary"
                  data-ouia-component-id="OUIA-Generated-Button-primary-3"
                  data-ouia-component-type="PF5/Button"
                  data-ouia-safe={true}
                  disabled={false}
                  onClick={[Function]}
                  role={null}
                  type="button"
                >
                  Create
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="resizable-ref"
      onMouseDown={[Function]}
    />
  </div>
</div>
`;

exports[`<JFRMetricsChartCard /> renders loading state correctly 1`] = `
<div
  className="draggable-ref-wrapper"
  onDragStart={[Function]}
  onTransitionEnd={[Function]}
  style={Object {}}
>
  <div
    className="dashboard-card-resizable-wrapper"
  >
    <div
      className="pf-v5-c-card pf-m-compact pf-m-rounded dashboard-card"
      data-ouia-component-id="OUIA-Generated-Card-2"
      data-ouia-component-type="PF5/Card"
      data-ouia-safe={true}
      id="CPU Load-chart-card"
      style={Object {}}
      title="CPU Load"
    >
      <div
        className="draggable-ref__grip"
        data-quickstart-id="card-draggable-grip"
        draggable={true}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="pf-v5-c-card__header"
        >
          <div
            className="pf-v5-c-card__actions"
          />
          <div
            className="pf-v5-c-card__header-main"
          >
            <div
              className="pf-v5-c-card__title"
            >
              <div
                className="pf-v5-c-card__title-text"
                id="CPU Load-chart-card-title"
              >
                CPU Load (last 120s, every 10s)
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="pf-v5-c-card__body"
      >
        <div
          className="pf-v5-l-bullseye"
        >
          <div
            className="pf-v5-c-empty-state"
          >
            <div
              className="pf-v5-c-empty-state__content"
            >
              <div
                className="pf-v5-c-empty-state__header"
              >
                <div
                  className="pf-v5-c-empty-state__icon"
                >
                  <svg
                    aria-hidden={false}
                    aria-label="Contents"
                    aria-valuetext="Loading..."
                    className="pf-v5-c-spinner pf-m-xl"
                    role="progressbar"
                    viewBox="0 0 100 100"
                  >
                    <circle
                      className="pf-v5-c-spinner__path"
                      cx="50"
                      cy="50"
                      fill="none"
                      r="45"
                    />
                  </svg>
                </div>
                <div
                  className="pf-v5-c-empty-state__title"
                >
                  <h2
                    className="pf-v5-c-empty-state__title-text"
                  >
                    Loading
                  </h2>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="resizable-ref"
      onMouseDown={[Function]}
    />
  </div>
</div>
`;
